### PR TITLE
HMACAuthenticationPolicy Unit test

### PR DESCRIPTION
**Pull request by JoshuaLai**
Setting up unit test to make sure the headers are being set as we expect.

### CODE REVIEW

File: sdk/communication/Azure.Communication.Common/tests/Pipeline/HMACAuthenticationPolicyTests.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+using Azure.Core;
+using Azure.Core.TestFramework;
+using NUnit.Framework;
+using NUnit.Framework.Internal;
+
+namespace Azure.Communication.Pipeline
+{
+    public class HMACAuthenticationPolicyTests : PolicyTestBase
+    {
+        private const string secret_key = "68810419818922fb0263dd6ee4b9c56537dbad914aa7324a119fce26778a286e";
```
Comment (RezaJooyandeh) [inline sdk/communication/Azure.Communication.Common/tests/Pipeline/HMACAuthenticationPolicyTests.cs:14]: nit: naming -> SecretKey I was surprised to see _ in the name of the original code. We do not normally use it in C#

File: sdk/communication/Azure.Communication.Common/tests/Pipeline/HMACAuthenticationPolicyTests.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+        [Test]
+        public async Task TestHMACPolicyProcess()
+        {
+            var shaValue = "47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU=";
+            var authPolicy = new HMACAuthenticationPolicy(secret_key);
+            var transport = new MockTransport(new MockResponse(200));
+            await SendGetRequest(transport, authPolicy);
```
Comment (RezaJooyandeh) [inline sdk/communication/Azure.Communication.Common/tests/Pipeline/HMACAuthenticationPolicyTests.cs:22]: I wonder if it makes sense to add similar methods with body that includes non English characters or the ones that would be escaped.

File: sdk/communication/Azure.Communication.Common/tests/Pipeline/HMACAuthenticationPolicyTests.cs [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            var headers = transport.SingleRequest.Headers;
+
+            Assert.True(headers.Contains("Date"));
+            Assert.True(headers.Contains("x-ms-content-sha256"));
+            Assert.True(headers.Contains("Authorization"));
+
+            foreach (HttpHeader h in headers) {
```
Comment (RezaJooyandeh) [inline sdk/communication/Azure.Communication.Common/tests/Pipeline/HMACAuthenticationPolicyTests.cs:29]: nit: `{` and `}` on a new line

File: sdk/communication/Azure.Communication.Common/tests/Pipeline/HMACAuthenticationPolicyTests.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+                if (h.Name == "x-ms-content-sha256") {
+                    Assert.AreEqual(h.Value, shaValue);
```
Comment (RezaJooyandeh) [inline sdk/communication/Azure.Communication.Common/tests/Pipeline/HMACAuthenticationPolicyTests.cs:31]: Any reason why not use the format of other examples:
```C#
Assert.True(headers.TryGetValue("x-ms-content-sha256", out string shaValue));
Assert.AreEqual(expectedShaValue, shaValue);
.... same for others
```

File: sdk/communication/Azure.Communication.Common/tests/Pipeline/PolicyTestBase.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Core.Pipeline;
+using NUnit.Framework;
+
+namespace Azure.Communication.Pipeline
+{
+    public abstract class PolicyTestBase
```
Comment (RezaJooyandeh) [inline sdk/communication/Azure.Communication.Common/tests/Pipeline/PolicyTestBase.cs:12]: We should be reuse the type from Azure core test, but worth inheriting from and adding overloads that we can pass a content to ensure we can handle none-ascii encoding and escaped characters (feel free to do the later in a separate PR)

File: sdk/communication/Azure.Communication.Common/tests/Pipeline/PolicyTestBase.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+            using (Request request = transport.CreateRequest())
+            {
+                request.Method = RequestMethod.Get;
+                request.Uri.Reset(new Uri("http://example.com"));
+                var pipeline = new HttpPipeline(transport, new[] { policy }, responseClassifier);
+                return await pipeline.SendRequestAsync(request, CancellationToken.None);
+            }
+        }
+
```
Comment (RezaJooyandeh) [inline sdk/communication/Azure.Communication.Common/tests/Pipeline/PolicyTestBase.cs:26]: nit: extra new line

**Review by RezaJooyandeh — COMMENTED** [review]
Thank you for adding the test 👏🏻. I added a couple of comments

File: sdk/communication/Azure.Communication.Common/tests/Pipeline/PolicyTestBase.cs (hunk 5 above)
Comment (JoshuaLai) [reply sdk/communication/Azure.Communication.Common/tests/Pipeline/PolicyTestBase.cs:12]: The original `PolicyTestBase` has an assert
`Assert.IsInstanceOf<HttpPipelineSynchronousPolicy>(policy, "Use SyncAsyncPolicyTestBase base type for non-sync policies");`
HMACAuthenticationPolicy didn't meet this requirement so I thought I would make our own version

File: sdk/communication/Azure.Communication.Common/tests/Pipeline/HMACAuthenticationPolicyTests.cs (hunk 4 above)
Comment (JoshuaLai) [reply sdk/communication/Azure.Communication.Common/tests/Pipeline/HMACAuthenticationPolicyTests.cs:31]: Oh I see I misunderstood how `TryGetValue` method works. I will definitely make this change

File: sdk/communication/Azure.Communication.Common/tests/Pipeline/HMACAuthenticationPolicyTests.cs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+            var expectedShaValue = "47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU=";
+            var authPolicy = new HMACAuthenticationPolicy(SecretKey);
+            var transport = new MockTransport(new MockResponse(200));
+            await SendGetRequest(transport, authPolicy);
+            var headers = transport.SingleRequest.Headers;
+
+            Assert.True(headers.Contains("Date"));
+            Assert.True(headers.Contains("Authorization"));
+
+            var shaValue = "";
```
Comment (RezaJooyandeh) [inline sdk/communication/Azure.Communication.Common/tests/Pipeline/HMACAuthenticationPolicyTests.cs:28]: You can inline the definition (`out var`):
```Assert.True(headers.TryGetValue("x-ms-content-sha256", out var shaValue));```

**Review by RezaJooyandeh — APPROVED** [review]

**Pull request merged**